plugins/outputs/postgresql/utils/utils.go

```diff
@@ -0,0 +1,203 @@
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"log"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/influxdata/telegraf"
+)
+
+const (
+	insertIntoSQLTemplate = "INSERT INTO %s(%s) VALUES(%s)"
+)
+
+func TagListToJSON(tagList []*telegraf.Tag) []byte {
+	tags := make(map[string]string, len(tagList))
+	for _, tag := range tagList {
+		tags[tag.Key] = tag.Value
+	}
+	bs, _ := json.Marshal(tags)
+	return bs
+}
+
+func FieldListToJSON(fieldList []*telegraf.Field) ([]byte, error) {
+	fields := make(map[string]interface{}, len(fieldList))
+	for _, field := range fieldList {
+		fields[field.Key] = field.Value
+	}
+	return json.Marshal(fields)
+}
+
+// QuoteIdent returns a sanitized string safe to use in SQL as an identifier
+func QuoteIdent(name string) string {
+	return pgx.Identifier{name}.Sanitize()
+}
+
+// QuoteLiteral returns a sanitized string safe to use in sql as a string literal
+func QuoteLiteral(name string) string {
+	return "'" + strings.Replace(name, "'", "''", -1) + "'"
+}
+
+// FullTableName returns a sanitized table name with it's schema (if supplied)
+func FullTableName(schema, name string) pgx.Identifier {
+	if schema != "" {
+		return pgx.Identifier{schema, name}
+	}
+
+	return pgx.Identifier{name}
+}
+
+// Constants for naming PostgreSQL data types both in
+// their short and long versions.
+const (
+	PgBool                     = "boolean"
+	PgSmallInt                 = "smallint"
+	PgInteger                  = "integer"
+	PgBigInt                   = "bigint"
+	PgReal                     = "real"
+	PgDoublePrecision          = "double precision"
+	PgNumeric                  = "numeric"
+	PgText                     = "text"
+	PgTimestampWithTimeZone    = "timestamp with time zone"
+	PgTimestampWithoutTimeZone = "timestamp without time zone"
+	PgSerial                   = "serial"
+	PgJSONb                    = "jsonb"
+)
+
+// DerivePgDatatype returns the appropriate PostgreSQL data type
+// that could hold the value.
+func DerivePgDatatype(value interface{}) PgDataType {
+	switch value.(type) {
+	case bool:
+		return PgBool
+	case uint64:
+		return PgNumeric
+	case int64, int, uint, uint32:
+		return PgBigInt
+	case int32:
+		return PgInteger
+	case int16, int8:
+		return PgSmallInt
+	case float64:
+		return PgDoublePrecision
+	case float32:
+		return PgReal
+	case string:
+		return PgText
+	case time.Time:
+		return PgTimestampWithTimeZone
+	default:
+		log.Printf("E! Unknown datatype %T(%v)", value, value)
```

Comment:
If you propagate the error upwards, you don't need the `log` facility.

Reply:
If we want to treat this as a non-recoverable error, we can. By logging we can try our best effort to add the data, and it may be successful.

If you want to make this non-recoverable, I think that's fine. I can't imagine why the plugin would receive a data type other that the ones that are already known/handled.